### PR TITLE
FEATURE: Implement Media Details Screen

### PR DESCRIPTION
**Pull request by grebaldi**
resolves: #73 

Screenshot_2022-11-05_13-27-24

This is a full integration of the Media Details Screen. It's available with `useNewMediaSelection` set to true. 

## `Interaction`: Ad-Hoc `confirm`-, `alert`- and `prompt`-style dialogs

This PR introduces the `Interaction` concept to the `@media-ui/core`. It is a context-driven API that needs to be initialized for every application entry point. The idea is to enable ad-hoc `confirm`-, `alert`- and `prompt`-style dialogs without much React ceremony, but instead with an API closer to the browser-native `confirm`, `alert` and `prompt` functions.

The call-site looks as simple as this:

```typescript
const SomeComponent = () => {
  const Interaction = useInteraction();

  const handleSomeAction = useCallback(
    async () => {
      if (await Interaction.confirm({ message: 'Do you really want to do this?' }) {
         await performSomeActionBecauseWeAreAllowedTo();
      }
    },
    [Interaction]
  );

  return (/* ... */);
};
```

Currently, only `confirm` has been implemented, because there's no need for the other ones in the scope of this PR (or perhaps the Media UI at large). In fact, this API (or one similar to this) should actually be provided by Neos.Neos.Ui.

## `ApprovalAttainmentStrategy`: To ask or not to ask?

This lengthy name refers to another concept introduced by this PR. The problem is that depending on the context certain actions (like updating asset properties) should require confirmation by the user. For example: While editing asset properties in the media module is not a problem at all, editing asset properties in the secondary inspector might give an editor the impression that the changes will only be applied to the image variant that is used as a node property.

**Media Module:**

https://user-images.githubusercontent.com/2522299/202213134-5782d745-9a7a-4b85-bd1c-164dc109f0c3.mp4

**Secondary Inspector:**

https://user-images.githubusercontent.com/2522299/202213303-fc51e071-1cd2-420a-8351-6eb764b3e9a7.mp4

To solve this, different application entry points can now provide an `ApprovalAttainmentStrategy`. Depending on the action an `ApprovalAttainmentStrategy` can use the `Interaction` API to show a confirmation dialog or just return `Promise.resolve(true)` to automatically give the approval.

Components like `<PropertyInspector />` then ask the currently applied `ApprovalAttainmentStrategy` for approval before they perform a critical action.

The `ApprovalAttainmentStrategy` covers the following actions:

- `updateAsset`
- `setAssetCollections`
- `setAssetTags`
- `deleteAsset`
- `replaceAsset`

By default, only `deleteAsset` will ask for confirmation (This also solves: `TODO: Use custom modal`, see: https://github.com/Flowpack/media-ui/blob/1326beff8efa46a38ae7872b53915015d35d9006/Resources/Private/JavaScript/core/src/provider/MediaUiProvider.tsx#L74).

The application entry point in `@media-ui/media-details-screen` adds confirmation dialogs for all actions.

## "Open in Media Ui" Action

https://user-images.githubusercontent.com/2522299/204830927-d6a36100-8f72-47d1-bf20-311d4a0d2df2.mp4

The Media Ui main application now reads the URL parameter `searchTerm`, which can be used as a means to initialize the search term input. Additionally, the search term can now have a special format `/id:([0-9a-f]{8}-[0-9a-f]{4}-[0-5][0-9a-f]{3}-[089ab][0-9a-f]{3}-[0-9a-f]{12})/` in which case the search will look for an asset by its Id rather than performing a full text search. Also, the asset passed via the search term will be selected automatically.

With both these features in place, it is now possible to create a link into the Media Ui main application with a preselected asset. The `<PreviewActions />`-Component of the media details screen now provides such a link. The Media Ui main application will respectively open in a separate tab and show the asset from media details screen.

Along with this feature come unit test setups for both Server-Side and Client-Side code (mainly to test the `SearchTerm`-related domain logic). Regarding the JavaScript tests, I'd like to address the elephant in the room right away:

### Why not `jest`?

I have chosen to use the good ol' `mocha` + `chai` combination for JavaScript unit tests. So, obviously, one might ask why not use `jest` (as probably the majority of contemporary projects do, including Neos.Ui)?

I recently have found `jest` to be a considerable maintenance burden with the node ecosystem moving towards native ESModule support. I have had projects halt the transition towards ESM entirely, because `jest` couldn't be made to work. ESM has been stable since Node v12 (see: https://nodejs.org/docs/latest-v12.x/api/esm.html), so for at least 3 years. TypeScript took a very long time to catch up to that, but since May of this year, full ESM support for Node has been added as well (see: https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/). Unfortunately, ESM support in `jest` is still an experimental feature (see: https://jestjs.io/docs/ecmascript-modules), so it is likely to still be a reason for many packages in the ecosystem to hold back their transition towards ESM. I'm aware that this is not an immediate problem for the Media Ui, since it's based on CommonJS packages. But since I have fought this problem multiple times already, I've decided to not start any new testing setups on the basis of `jest`, but use `mocha` + `chai` instead, because both these packages are well maintained and just work.

`mocha` + `chai` do not come without their own problems though. Notably, I had to exclude `@types/mocha` from the setup, because those types were colliding with globals registered by `testcafe` (this is being addressed here: https://github.com/DevExpress/testcafe/pull/6338). However, it is likely this issue would have come up with `jest` as well.

All of this being said, I wish to point out, that I'm of course open to discuss this point (This also applies to every other implementation decision I have made in this PR :)). 

### TODOs

- [x] The `/id:([0-9a-f]{8}-[0-9a-f]{4}-[0-5][0-9a-f]{3}-[089ab][0-9a-f]{3}-[0-9a-f]{12})/` format is likely to be a false assumption, because assets from foreign asset sources may have Ids that are not UUIDs.
  - This has been worked out by @Sebobo. By using the asset source `neos` to retrieve the local asset, Ids are guaranteed to be UUIDs.
- [x] None of this works with imported assets right now, because the image editor (Neos.Ui) only passes the local asset id rather than asst id + source id.
  - Same here: Using the asset source `neos` solves this.

## TODO

- [x] Move MediaDetailsScreen into separate feature package `@media-ui/media-details-screen`
- [x] Create `@media-ui/neos-ui-plugin` package and move plugin integration logic there (for both details screen and selection screen)
- [x] AssetInspector Integration
  - [x] <s>Make "Replace Asset" work</s>, OR: Remove "replace" button?
    - *Note:* While "Replace Asset" works in principle, we won't be able to update the image in the inspector preview for now. For initial release it makes sense to remove the "replace" button.
- [x] Toolbar Integration
  - <s>Make "Preview Asset" work</s>
  - [x] Replace "Preview Asset" with "Open asset in new tab"
  - <s>Make "Delete Asset" work</s>
  - [x] Remove "Delete Asset" action (see comment: https://github.com/Flowpack/media-ui/pull/148#issuecomment-1317166825)
- <s>Show Title, description & copyright in AssetEditor (see Shel.imageValidator) hidden under a button with an "I" icon</s>
  - <s>Allow adding rules like in the validator package and show an alert if they are not fulfilled</s>
  - *Not part of this issue, see comment: https://github.com/Flowpack/media-ui/pull/148#issuecomment-1309936936*
- [x] Move asset inspector to the left, asset preview right, 50%/50% layout probably
  - For now, I decided to give only `theme.sidebarWidth` space for the asset inspector, to leave more room for the asset itself
- [x] When editing and saving changes a confirm popup should inform the user that the changes are done globally
- [x] Find a way to go from secondary inspector to the media module with the current asset selected
- [x] Allow downloading an asset via a button
- [x] the screen pops while loading the asset due to the large placeholder. So we need a loading indicator here which is not so intense.
- [ ] the inspector is not scrollable for me and not all information can be reached
- [x] switching to the variants tab breaks the layout height and the asset preview disappears
- [x] link styling in the usage dialog is incorrect (dark blue)
- [x] checkbox styling in replace dialog is broken (line break and missing icons)
- [x] Update `@neos-project/react-ui-components` to the latest `5.3.*`
  - *Note:* Latest would be `5.3.18` atm, but I went for `5.3.13` instead, because the package patch for `@neos-project/react-ui-components` won't work otherwise and `5.3.13` contains https://github.com/neos/neos-ui/commit/abceeccde9ba5797b88656bd921095c67f620f5d, which is the change that needed to be addressed.

### CODE REVIEW

**Comment by Sebobo** [comment]
Hi @grebaldi, thx for jumping on this topic! 
Did you see the results of our related discussion during the sprint in #146 ?

Especially with actions like replace, delete etc. we have to be careful in the UX as we need to make sure an editor understands the global consequences.

I wouldn't introduce a separate flag for this as we can replace all old features and I don't expect issues there. 
And a separate feature package would be good.

**Comment by Sebobo** [comment]
> Thanks for the pointer, I wasn't aware of the discussion :)
👍 no problem, should have linked the issue
 
> > Show Title, description & copyright in AssetEditor (see Shel.imageValidator) [...]
> 
> Does this refer to [Shel.Neos.ImageChecker](https://github.com/Sebobo/Shel.Neos.ImageChecker)?

Yes, but it's isn't part of this issue. As the rewrite of the AssetEditor is separate (big) topic.
 
> I'd prefer the second approach. What do you think, @Sebobo?

👍 yes that sounds perfect. It might be a bit confusing now that they all lie next to each other but have those not super obvious interdependencies. So at some point we might need to have 3 folders for packages, ui plugin, backend module and shared stuff.

**Comment by grebaldi** [comment]
I think there's a way to make `replaceAsset` work in the secondary inspector. But `deleteAsset` doesn't make much sense there, because one would get `Exception #1462196420:  Asset could not be deleted, because it is still in use.` anyway - which means the asset simply cannot be deleted if it's available to the secondary inspector.

I'm going to experiment a little to see, whether `replaceAsset` can be done.

To hide `deleteAsset` (and potentially `replaceAsset` as well), I think it would be best to use the existing feature flag mechanism instead of adding more context. The application entry point in `@media-ui/media-details-screen` would then be able to override the global feature flag.

wdyt, @Sebobo?

**Comment by Sebobo** [comment]
`deleteAsset` makes no sense, I agree. So just get rid of it.
If we have a helpful dialog for `replaceAsset` I'm fine without a feature flag.

**Comment by Sebobo** [comment]
Hi @grebaldi I finally got the devserver and  tests working again in #154 
You should probably rebase even though the Details Screen is only in the ui plugin.

I also created a new issue #157 to start the process to also run e2e tests for the ui plugin.

**Comment by grebaldi** [comment]
Hi @Sebobo,

I've rebased and everything is green :tada: 

I'm currently stuck with a problem that we should discuss probably: The image editor (Neos.Ui) passes an `assetIdentity` prop to the media details screen. For imported assets, this `assetIdentity` contains the local asset Id. This local Id cannot be resolved atm, so the details screen fails with imported assets.

I suppose it is possible to resolve the local asset Id, but I think we should discuss how to go about this.

**Comment by Sebobo** [comment]
I think for now we should disable the button then if we have an imported asset. There are several issues with imported assets atm and I have to workaround them in other places too. 

So we should discuss this separately as the Media API needs to be fixed for that. But targeting this for Neos 9 would be essential I think.

**Comment by Sebobo** [comment]
Can't get your branch to work properly right now:

Bildschirm­foto 2022-12-02 um 13 50 34

There are not console errors, so I have to dig deeper when I have more time.

**Comment by grebaldi** [comment]
@Sebobo 
> There are not console errors, so I have to dig deeper when I have more time.

Never mind, I'll take a look. Probably missed a bug with the recent changes, because I tested with imported assets last (and the screen you're seeing is the current behavior for imported assets - I apparently overlooked that regular assets are broken too).

> I think for now we should disable the button then if we have an imported asset.

I believe that would require a change for the Image Editor (Neos.Ui). Regarding that: I've been thinking that it might be sensible to move that editor entirely over to  Flowpack.Media.Ui (it definitely makes more sense to let it reside closer to the Media Context rather than the Core).

Such a migration would be out of scope for this PR though. So, I'm considering two options for going forward:

1. **Either:** Leave imported assets broken in the details screen for now, but merge (and ship) anyway if everything else works.
2. **Or:** Open up a long-running branch from `master`, move this PR over there and use the new branch to migrate the Image Editor, do imported assets fixes and so on. Finally, merge the long-running branch back into `master` once everything works.

Wdyt?

**Comment by grebaldi** [comment]
@Sebobo 
> @grebaldi: Never mind, I'll take a look. Probably missed a bug with the recent changes, because I tested with imported assets last (and the screen you're seeing is the current behavior for imported assets - I apparently overlooked that regular assets are broken too).

I investigated locally and regular assets seem to work fine for me. So, we'd need to take a look at what's happening in your setup.

File: Resources/Private/JavaScript/media-details-screen/src/MediaDetailsScreen.tsx [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+                                        approvalAttainmentStrategyFactory={
+                                            MediaDetailsScreenApprovalAttainmentStrategyFactory
+                                        }
+                                    >
+                                        <MediaUiThemeProvider>
+                                            <Details
+                                                buildLinkToMediaUi={buildLinkToMediaUi}
+                                                assetIdentity={{
+                                                    assetId: this.props.imageIdentity,
+                                                    assetSourceId: '',
```
[the remark this answers is not in the document]
Comment (Sebobo) [reply Resources/Private/JavaScript/media-details-screen/src/MediaDetailsScreen.tsx:171]: Interestingly enough it still works with assets from other sources 🤷🏻‍♂️

**Comment by grebaldi** [comment]
Hi @Sebobo,

I was able to fix some of the issues you have found (and I'll continue with the rest of course). Along the way there are a couple of questions/remarks I'd like to clarify:

> checkbox styling in replace dialog is broken (line break and missing icons)

I was able to fix the line break, but the missing icons seem to be an issue with Neos UI itself. In plugin context, the dialog uses the Checkboxes as provided by Neos UI. Those Checkboxes do not provide the greyed-out checkmark icon (maybe an older version did?).

> the inspector is not scrollable for me and not all information can be reached

What browser did you use for testing? I was not able to reproduce this issue in Gecko, WebKit or Blink on Linux.

> switching to the variants tab breaks the layout height and the asset preview disappears

How do I get to the variants tab? (Full disclosure: I don't know what the variants tab is :sweat_smile:)

**Comment by Sebobo** [comment]
Thx!

The variant tab is enabled via the feature flag `Neos.Neos.Ui.frontendConfiguration.Flowpack.Media.Ui.showVariantsEditor`.

I'm testing with Brave. Will check your latest version later today and see if the error is still there and if I see why it's there if so.

Regarding the icon we then can check whether we need a hot fix in our code or fix it in the UI itself.

**Comment by grebaldi** [comment]
Looks like those icons have been explicitly removed: https://github.com/neos/neos-ui/pull/2995

The `media-module` package currently requires:
```json
"@neos-project/react-ui-components": "5.3.7"
```

The commit on Neos UI PR 2995 seems only to be included as early as `5.3.13`:
https://github.com/neos/neos-ui/commit/abceeccde9ba5797b88656bd921095c67f620f5d

That explains why the icon shows up in the backend module but not in Neos UI.

For a fix, I could update `@neos-project/react-ui-components` to the latest `5.3.*`, but I seem to recall that there was an issue with updating the components, right?

**Comment by Sebobo** [comment]
ah ok 👌 Updating the components should be no problem. We have some other dependencies which break everything when being updated.

**Comment by Sebobo** [comment]
Looks good so far. My first tests looked fine!

I still have the issue with the content exceeding the content area and breaking the ui layout. 
It might have something todo with the CSS Grid I introduced to the UI in 8.2. 
Which Neos version are you using?

**Comment by grebaldi** [comment]
@Sebobo 
> I still have the issue with the content exceeding the content area and breaking the ui layout.
> It might have something todo with the CSS Grid I introduced to the UI in 8.2.
> Which Neos version are you using?

Yeah, this was it :) I've started development on 8.1 and didn't upgrade since. I can reproduce the issue now on 8.2 and will come up with fix soonish.

**Review by Sebobo — APPROVED** [review]

**Comment by Sebobo** [comment]
Big ❤️, thx Wilhelm

**Pull request merged**